pystachio/choice.py

```diff
@@ -0,0 +1,166 @@
+# Choice types: types that can take one of a group of selected types.
+
+import json
+
+from .base import Object
+from .basic import SimpleObject
+from .compatibility import Compatibility
+from .typing import (
+    Type,
+    TypeCheck,
+    TypeFactory,
+    TypeMetaclass
+)
+
+
+class ChoiceFactory(TypeFactory):
+    """A Pystachio type representing a value which can be one of several
+    different types.
+    For example, a field which could be either an integer, or an integer
+    expression (where IntegerExpression is a struct type) could be written
+    Choice("IntOrExpr", (Integer, IntegerExpression))
+    """
+    PROVIDES = 'Choice'
+
+    @staticmethod
+    def create(type_dict, *type_parameters):
+        """
+        type_parameters should be:
+            (name, (alternative1, alternative2, ...))
+        where name is a string, and the alternatives are all valid serialized
+        types.
+        """
+        assert len(type_parameters) == 2
+        name = type_parameters[0]
+        alternatives = type_parameters[1]
+        assert isinstance(name, Compatibility.stringy)
+        assert isinstance(alternatives, (list, tuple))
+        choice_types = []
+        for c in alternatives:
+            choice_types.append(TypeFactory.new(type_dict, *c))
+        return TypeMetaclass(str(name), (ChoiceContainer,), {'CHOICES': choice_types})
+
+
+class ChoiceContainer(Object, Type):
+    """The inner implementation of a choice type value.
+
+    This just stores a value, and then tries to coerce it into one of the alternatives when
+    it's checked or interpolated.
+    """
+    __slots__ = ('_value',)
+
+    def __init__(self, val):
+        self._value = val
+        self._scopes = ()
+
+    def scopes(self):
+        return self._scopes
+
+    def get(self):
+        if isinstance(self._value, Object):
+            return self._value.get()
+        else:
+            return self._value
+
+    def value(self):
```

Comment:
.get() returns the tuple-ized primitive value. This returns whatever value is wrapped in the Pystachio choice object. There are places where you want to look at that. 

In code that uses configs, we have a couple of places where we want to look at a choice value, and be able to ask "Which option is this?". It's easier to do that with the Pystachio value than with the tuple returned by get.